lib/ansible/modules/windows/win_firewall_rule.ps1

```diff
@@ -59,346 +77,380 @@ function preprocessAndCompare($key, $outputValue, $fwsettingValue) {
-        $result = @{
-            failed = $false
-            exists = $exists
-            identical = $correct
-            multiple = $multi
-            difference = $difference
-            msg = $msg
-        }
-    } catch [Exception]{
-        $result = @{
-            failed = $true
-            error = $_.Exception.Message
-            msg = $msg
+            $result.msg += @("No rule '$name' could be found")
         }
-    };
+    } catch [Exception] {
+        $result.failed = $true
+        $result.error = $_.Exception.Message
+    }
     return $result
-};
+}
 
 function createFireWallRule ($fwsettings) {
-    $msg=@()
-    $execString="netsh advfirewall firewall add rule"
+    $result = @{
+        changed = $false
+        diff = @{}
+        failed = $false
+        msg = @()
+    }
 
+    $command = "netsh advfirewall firewall add rule"
     ForEach ($fwsetting in $fwsettings.GetEnumerator()) {
-        if ($fwsetting.key -eq 'Direction') {
-            $key='dir'
-        } elseif ($fwsetting.key -eq 'Rule Name') {
-            $key='name'
-        } elseif ($fwsetting.key -eq 'Enabled') {
-            $key='enable'
-        } elseif ($fwsetting.key -eq 'Profiles') {
-            $key='profile'
-        } else {
-            $key=$($fwsetting.key).ToLower()
-        };
-        $execString+=" ";
-        $execString+=$key;
-        $execString+="=";
-        $execString+='"';
-        $execString+=$fwsetting.value;
-        $execString+='"';
-    };
-    try {
-        #$msg+=@($execString);
-        $output=$(Invoke-Expression $execString| ? {$_});
-        $msg+=@("Created firewall rule $name");
-
-        $result=@{
-            failed = $false
-            output=$output
-            changed=$true
-            msg=$msg
-        };
+        if ($fwsetting.value -ne $null) {
+            switch($fwsetting.key) {
+                "Direction" { $option = "dir" }
+                "Rule Name" { $option = "name" }
+                "Enabled" { $option = "enable" }
+                "Profiles" { $option = "profile" }
+                "InterfaceTypes" { $option = "interfacetype" }
+                "Security" { $option = "security" }
+                "Edge traversal" { $option = "edge" }
+                default { $option = $($fwsetting.key).ToLower() }
+            }
+            $command += " $option='$($fwsetting.value)'"
+        }
+    }
 
+    try {
+        $rc = 0
+        if (-not $check_mode) {
+            $result.output = Invoke-Expression $command | Where { $_ }
+            $rc = $LASTEXITCODE
+        }
+        if ($rc -eq 0) {
+            if ($diff_support) {
+                $result.diff.after = ConvertTo-SortedKV $fwsettings
+                $result.diff.before= ""
+            }
+            $result.changed = $true
+            $result.msg += @("Created firewall rule '$name'")
+        } else {
+            $result.failed = $true
+            $result.msg += @("Create command '$command' failed with rc=$rc")
```

Comment:
I'm a) new to PowerShell b) to the Ansible internals and c) Windows so disregard if needed but will this retain the error message somehow that netsh returned? That's helpful for debugging.